src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfo.cs

```diff
@@ -250,6 +250,84 @@ private static double GetNumericValueNoBoundsCheck(uint codePoint)
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToUpper(char codePoint)
+        {
+            nuint offset = GetCategoryCasingTableOffsetNoBoundsChecks((uint)codePoint);
+
+            // The offset is specified in shorts:
+            // Get the 'ref short' corresponding to where the addend is, read it as a signed 16-bit value, then add
+
+            ref short rsStart = ref Unsafe.As<byte, short>(ref MemoryMarshal.GetReference(UppercaseValues));
+            ref short rsDelta = ref Unsafe.Add(ref rsStart, (nint)offset);
+            nint delta = (BitConverter.IsLittleEndian) ? rsDelta : BinaryPrimitives.ReverseEndianness(rsDelta);
+            return (char)(delta + (nint)codePoint);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static uint ToUpper(uint codePoint)
+        {
+            if (UnicodeUtility.IsBmpCodePoint(codePoint))
+            {
+                return ToUpper((char) codePoint);
+            }
+
+            if (!UnicodeUtility.IsValidCodePoint(codePoint))
+            {
+                // We don't throw here even if we are getting invalid codepoint. We just return the same input value at that time.
+                return codePoint;
+            }
+
+            nuint offset = GetCategoryCasingTableOffsetNoBoundsChecks(codePoint);
+
+            // The offset is specified in shorts:
+            // Get the 'ref short' corresponding to where the addend is, read it as a signed 16-bit value, then add
+
+            ref short rsStart = ref Unsafe.As<byte, short>(ref MemoryMarshal.GetReference(UppercaseValues));
+            ref short rsDelta = ref Unsafe.Add(ref rsStart, (nint)offset);
+            nint delta = (BitConverter.IsLittleEndian) ? rsDelta : BinaryPrimitives.ReverseEndianness(rsDelta);
+            return (uint)(delta + (nint)codePoint);
```

Comment:
Should this calculation be a helper method? Or it messes up the inlining?

Reply:
I don't think it is worth to move it to its own method.